src/hypertable_restrict_info.c

```diff
@@ -0,0 +1,358 @@
+{
+	int64		value = partitioning_func_apply(dri->base.dimension->partitioning, c->constvalue);
+
+	switch (strategy)
+	{
+		case BTEqualStrategyNumber:
+			dri->value = value;
+			dri->strategy = strategy;
+			return true;
+		default:
+			return false;
+	}
+}
+
+
+static bool
+dimension_restrict_info_add(DimensionRestrictInfo *dri, int strategy, Const *c)
+{
+	switch (dri->dimension->type)
+	{
+		case DIMENSION_TYPE_OPEN:
+			return dimension_restrict_info_open_add((DimensionRestrictInfoOpen *) dri, strategy, c);
+		case DIMENSION_TYPE_CLOSED:
+			return dimension_restrict_info_closed_add((DimensionRestrictInfoClosed *) dri, strategy, c);
+		default:
+			elog(ERROR, "unknown dimension type");
+	}
+}
+
+static DimensionVec *
+dimension_restrict_info_open_slices(DimensionRestrictInfoOpen *dri)
+{
+	/* basic idea: slice_end > lower_bound && slice_start < upper_bound */
+	return dimension_slice_scan_range_limit(dri->base.dimension->fd.id, dri->upper_strategy, dri->upper_bound, dri->lower_strategy, dri->lower_bound, 0);
+}
+
+static DimensionVec *
+dimension_restrict_info_closed_slices(DimensionRestrictInfoClosed *dri)
+{
+	if (dri->strategy == BTEqualStrategyNumber)
+	{
+		/* slice_end >= value && slice_start <= value */
+		return dimension_slice_scan_range_limit(dri->base.dimension->fd.id,
+												BTLessEqualStrategyNumber,
+												dri->value,
+												BTGreaterEqualStrategyNumber,
+												dri->value,
+												0);
+	}
+	else
+	{
+		/* get all slices */
+		return dimension_slice_scan_range_limit(dri->base.dimension->fd.id,
+												InvalidStrategy,
+												-1,
+												InvalidStrategy,
+												-1,
+												0);
+	}
+}
+
+static DimensionVec *
+dimension_restrict_info_slices(DimensionRestrictInfo *dri)
+{
+	switch (dri->dimension->type)
+	{
+		case DIMENSION_TYPE_OPEN:
+			return dimension_restrict_info_open_slices((DimensionRestrictInfoOpen *) dri);
+		case DIMENSION_TYPE_CLOSED:
+			return dimension_restrict_info_closed_slices((DimensionRestrictInfoClosed *) dri);
+		default:
+			elog(ERROR, "unknown dimension type");
+	}
+}
+
+typedef struct HypertableRestrictInfo
+{
+	int16		max_attr;
+	int			num_base_restrictions;	/* number of base restrictions
+										 * successfully added */
+	DimensionRestrictInfo *dimension_restriction[FLEXIBLE_ARRAY_MEMBER];	/* Sparse. key is
+																			 * attrOffset of the
+																			 * dimension */
+} HypertableRestrictInfo;
+
+
+HypertableRestrictInfo *
+hypertable_restrict_info_create(RelOptInfo *rel, Hypertable *ht)
+{
+	HypertableRestrictInfo *res = palloc0(sizeof(HypertableRestrictInfo) + sizeof(DimensionRestrictInfo *) * rel->max_attr);
+	int			i;
+
+	res->max_attr = rel->max_attr;
+	for (i = 0; i < ht->space->num_dimensions; i++)
+	{
+		DimensionRestrictInfo *dri = dimension_restrict_info_create(&ht->space->dimensions[i]);
+
+		res->dimension_restriction[AttrNumberGetAttrOffset(ht->space->dimensions[i].column_attno)] = dri;
```

Comment:
While I agree a list would probably not be /bad/ I think the sparse array is more efficient because of O(1). Since we may have many clauses, I'm not sure why we wouldn't use this. The memory usage is limited as I mentioned before.

Reply:
There's really only a benefit of O(1) lookups when you have big data sets and not with one or two elements, which is the common case here. I mean, honestly, most of the time your are creating a sparse array with one single element! (Or, am I missing something?). This seems like over-engineering of an otherwise very simple thing. I wouldn't push back if you had a strong argument here, like showing an important efficiency improvement (e.g., significantly faster planning times). But I think, when in doubt, we should go for simplicity and maintainability of the code with the option of optimizing in the future. 

Since this seems like a "won't fix", I guess you strongly believe this is an important efficiency/speed optimization, to the extent that it is worth pushing it through. Thus I won't block the PR on this.

Reply:
Fixed - made it into a non-sparse array